nmea_parser.c

```diff
@@ -1,191 +1,315 @@
-						// volume up
-						debug_print("Volume up\n");
-						change_volume(+10.0);
-					}
-					if (*(ptr+1) == 'D') {
-						// volume down
-						debug_print("Volume down\n");
-						change_volume(-10.0);
-					}
-					if (*(ptr+1) == 'M') {
-						// Toggle Mute
-						debug_print("Toggle Mute\n");
-						toggle_mute();
-					}
-					if (*(ptr+1) == 'A' &&  *(ptr+2) == 'R') {
-						// Set Vario Mode
-						debug_print("Set Vario Mode\n");
-						set_vario_mode(vario);
+	buffer[2000]=0;
+	if ((ptr2=strchr(buffer,'$'))!=NULL) {
+		for (i=0;ptr2[i]!=0;++i) buffer[i]=ptr2[i];
+		buffer[i]=0;
+	}
+	if (strlen(buffer)+strlen(message)>1999) 
+		strcpy(buffer,message);
+	else strcpy(buffer+strlen(buffer),message); // Append the new sentence to the last
+	ptr = strtok(buffer, del1); // Tokenize about '$'
+	last=buffer;
+	while (ptr != NULL) {
+		last=ptr;
+		tick=1;
+		if ((*ptr=='P') && (*(ptr+1)=='O') && (*(ptr+2)=='V') && (*(ptr+3)==',')) { // If you begin the sentence correctly  
+			if (verify_checksum(ptr)) { // And checksum passes
+				tick=0;
+				ptr2=ptr+4; // Start looking after the $POV
+				ptr+=strlen(ptr)+1; // Move ptr to the next POV
+				ptr2=strtok(ptr2,del2); // Tokenize about ','
+				while (ptr2 != NULL) { // Repeat until you hit the next '$'  
+					switch (*ptr2) {
+						case 'C':
+							// Command sentence
+							// get next value
+							ptr2 = strtok(NULL, del2);
+							switch (*ptr2) {
+								case 'M':
+									if (*(ptr2+1) == 'C') {
+										//Set McCready
+										status&=(0x04^0xff);
+										ptr2 = strtok(NULL, del2);
+										setMC(atof(ptr2));
+										debug_print("Get McCready Value: %f\n",atof(ptr2));
+									}
+									break;
+								case 'W':
+									if (*(ptr2+1) == 'L') {
+										//Set Wingload/Ballast
+										status&=(0x08^0xff);
+										ptr2 = strtok(NULL, del2);
+										setBallast(atof(ptr2));
+										debug_print("Get Ballast Value: %f\n",atof(ptr2));
+									}
+									break;
+								case 'B':
+									if (*(ptr2+1) == 'U') {
+										//Set Bugs
+										status&=(0x10^0xff);
+							 			ptr2 = strtok(NULL, del2);
+										setDegradation(atof(ptr2));
+										debug_print("Get Bugs Value: %f\n",atof(ptr2));
+									}
+									break;
+								case 'P':
+									//Set Polar
+									// depricated
+									if (*(ptr2+1) == 'O' &&  *(ptr2+2) == 'L') {
+										ptr2 = strtok(NULL, del2);
+										polar.a=atof(ptr2);
+										ptr2 = strtok(NULL, del2);
+										polar.b=atof(ptr2);
+										ptr2 = strtok(NULL, del2);
+										polar.c=atof(ptr2);
+										ptr2 = strtok(NULL, del2);
+										polar.w=atof(ptr2);
+										setPolar(polar.a, polar.b, polar.c, polar.w);
+										status&=(0x01^0xff);
+									}
+									break;
+								case 'R':
+									//Set Real Polar
+									printf ("%s\n",ptr2);
+									if (*(ptr2+1) == 'P' &&  *(ptr2+2) == 'O') {
+										if (read_float_from_sentence(3,fvals,ptr2+4,del2)) {
```

Comment:
You should use this function in other places, too.
Add the end_of_sentence pointer to the function and you can do all the necessary checks inside the function in a consistent way.